### ISSUE TITLE
Native support for `TRY/EXCEPT` functionality

### ISSUE THREAD

**Issue by pekkaklarck** [robotframework/robotframework#3075]
The plan is to add support for native `IF/ELSE` syntax in RF ~3.3~ 4.0. See the discussion in #3074 for background information why we think adding this kind of programmatic features is a good idea. Another similar and commonly needed feature is `TRY/EXCEPT` functionality to ease handling errors. Currently it is possible to use `Run Keyword And Expect Error` and `Run Keyword And Ignore` error but they are awkward similarly as `Run Keyword If` is an awkward replacement for `IF/ELSE` syntax.

Exact semantics how `TRY/EXCEPT` will work are not yet clear. They will be decided later before implementation starts. A precondition is a new parser, similarly as for `IF/ELSE`, and we also consider `IF/ELSE` more important and will start implementing it first.

**Comment by Snooz82** [comment]
Just an idea:

At the moment we have already TRY/FINALLY with teststeps and Teardown.
Teardown can be on every level.

What is missing is only a [On-Error] setting.
If we would have this [On-Error] we could also do stuff like capture screenshots there and closing connections an so on.

I personally would not like to see a real TRY/EXCEPT like IF/ELSE, because the we start re-implementing Python.
[On-Error] is also understandable for non programmers.

```
*** Test Cases ***
Test
    [Setup]     Open Browser.   http://example.com
    Do Keyword
    Some Other Test
    [On-Error]    Save error log and capture Screenshot
    [Teardown]    Close All Browsers
```

Same could work on Suite and on Keyword Level.
```
*** Settings *** 
Suite Setup   Open Browser
Error Handler    Save error log and capture Screenshot
Suite Teardown    Close All Browsers
```

**Comment by pekkaklarck** [comment]
I've been thinking about adding an on-error handler but nowadays I think TRY/EXCEPT would probably be better. Some of the problems I see with on-error handler:

- Unless on-error can also ignore the error, it's not really different from teardown with `Run Keyword If Test Failed`.
- If on-error should be able to ignore errors, it needed to get exception information as argument to be able to do decisions based on that.
- One test (or keyword) could only have one on-error handler. If there were multiple possible errors to handle, the handler would get complicated.

I got a feeling TRY/EXCEPT would be a lot more generic solution. In simple cases on-error might be easier to understand, but in more complex scenarios TRY/EXCEPT would be a lot more flexible.

I also don't see problems in Robot's syntax moving towards a general very high level programming language. After replacing Run Keyword If with IF/ELSE and Run Keyword And Expect/Ignore Error with TRY/EXCEPT, we can look at replacing Exit/Continue For Loop with BREAK/CONTINUE, [Return] and Return From Keyword with RETURN, and so on.

**Comment by mkorpela** [comment]
How does this relate to expecting an error to happen? In otherwords many cases in tests the situation is so that an error must happen as expected, test should fail if not.

**Comment by pekkaklarck** [comment]
We could support `TRY/EXPECT/ELSE` like Python to support that use case.

**Comment by mkorpela** [comment]
One of the strongest selling points of Robot Framework is the simplicity for people coming from a non-IT-background. I think we should consider how easy a new core concept is to learn and how easy it is to teach.
TRY/EXCEPT goes some steps towards more complex concepts.

**Comment by Snooz82** [comment]
I do see usecases for Try/Except but these are mostly better done in python.

I always hate it when i see testspecs with try catch in it. People try to implement execution logic into test design.

@pekkaklarck you mentioned the "Run Keyword If Test Failed" this imho does not work on keyword level. But there are imho regular cases where you want to clean up or do stuff if something fails. And  using TRY EXCEPT for this would imho be too technical. 🤷🏻‍♂️

**Comment by aikarjal** [comment]
+1 for this feature. There are valid use-cases for this in RPA. Right now the only way to handle errors is awkward `Run Keyword And Expect Error`

**Comment by kariharju** [comment]
Definitely coming more and more of an issue, so nice to see this as critical v5 content.

**Comment by pekkaklarck** [comment]
This feature has been developed [in a branch](https://github.com/robotframework/robotframework/tree/try-except) and is getting ready to be merged. The basic functionality works like this:

```
TRY
    Some keyword
EXCEPT    Error message
    Error handler keyword
END
```

If `Some keyword` passes, the `EXCEPT` branch isn't run. If it fails with message `Error message`, the `EXCEPT` branch is executed. If it fails any other way, error is not handled. Any branch could naturally have more than one keyword.

There can be more than one `EXCEPT` branches. In that case they are matched one by one and the first one matching is executed. One `EXCEPT` can also have multiple messages to match. If it has no messages to match, it matches any:

```
TRY
    Some keyword
EXCEPT    Error message
    Error handler 1
EXCEPT    Alternative error
    Error handler 2
EXCEPT    Another error    One more error    # Match either of these
    Error handler 3
EXCEPT    # Match anything
    Error handler 3
END
```

By default matching the error requires an exact match. That can be changed by prefixing the message with `glob:`, `regexp:` or `starts:` to make the match a glob pattern match, regular expression match, or to match only the beginning of the error, respectively. If with patterns you are interested in the actual error message, you can assign it to a variable similarly as you can use `except Exception as err` in Python:

```
TRY
    Some keyword
EXCEPT    glob: Error *    AS    ${err}
    Log to console    Ignoring error: ${err}
END
```

Assigning to a variable works also if you don't specify the message at all i.e. when all messages match:

```
TRY
    Some keyword
EXCEPT    AS    ${message}
    Log to console    Error occurred: ${message}
END
```

It's possible to handle the case when there's no error using an `ELSE` branch like in Python:

```
TRY
    Some keyword
EXCEPT    Message
    Error handler keyword
ELSE
    Log to console    No error!
END
```

Using `ELSE` without any `EXCEPT` branches is a syntax error.

Finally, it's possible to execute code both when there's an error and when there isn't by using a `FINALLY` branch:

```
TRY
    Some keyword
EXCEPT    AS    ${err}
    Log to console    Error occurred: ${err}
ELSE
    Log to console    No error occurred.
FINALLY
    Log to console    Finally we are in FINALLY.
END
```

It is allowed to use `FINALLY` without `EXCEPT` branches. Such `TRY/FINALLY` structures can be used for cleaning up resources regardless are there failures or not:

```
Open connection
TRY
    Use connection
FINALLY
    Close connection
END
```    

Although this is mostly ready, there are still some design decisions that are still somewhat open. I'm going to write about them in separate comments.

**Comment by pekkaklarck** [comment]
When matching error messages, we have decided to require an exact match by default. That's different compared to `Run Keyword And Expect Error` that used glob patterns by default. A problem with glob patterns is that if the error contains meta-characters `*`, `?` or `[]` results can be strange. This is the reason `Run Keyword And Expect Error` got support for other matching approaches in RF 3.1 (#2967). With `TRY/EXCEPT` we decided to use safer exact match approach instead and allow same matching configuration as `Run Keyword And Expect Error` has. One reason we consider requiring an exact match by default is safe is that using `EXCEPT` without any patterns matches anything so there's no need for something like `glob:*`.

Although we somewhat firmly believe requiring an exact match by default is the best approach, this design decision is something we can still consider to change if there are good reasons to do so. If you like the current behavior, add :+1: to this comment and in other case add :-1:. Especially in the latter case, please add a comment why you thing the behavior should change.

**Comment by pekkaklarck** [comment]
Related to matching, we could also consider adding other matching strategies than current "equals", "glob", "regexp" and "starts". Things that could be added include "contains" and somehow making matching case-insensitive.

Matching exceptions by their type like `type:ValueError` could also be convenient. It ought to be relatively easy to implement if library keywords are used directly in the `TRY` branch, but handling possibly deeply nested user keywords could be more complicated. Even more problematic would be situation where there are multiple failures due to continuable failures or because there is failure both in a user keyword body and in its teardown. Probably in these cases "type" matching could simply always fail.

Add a comment if you think any new matching approach would be useful! They can also be added in future versions based on concrete needs after we get more experience from this new functionality. Adding new prefixes like `type:` isn't fully backwards compatible but it's very unlikely someone would actually have a message starting like that.

**Comment by pekkaklarck** [comment]
One thing to decide is what the variable created when using the `AS  ${err}` syntax should actually contain. Currently it contains the error message and that's also what `Run Keyword And Expect Error` returns.

Alternatively `${err}` could be the actual exception that occurred. It would allow more advanced inspection of the occurred error, but it would also make the common case of inspecting or just logging the error message more complicated. For non-technical users suddenly needing to deal with Python exception objects could be really complicated in general.

Yet another approach would be making `${err}` our custom object that contains the error message (e.g. as `${err.message}`) but also has a reference to the actual exception (e.g. `${err.exception}`) and possibly also other information like formatted traceback. Also this would make the most common case of just using the message for something a bit more complicated than the current behavior, but the possibility to inspect the actual exception could be worth that.

We were actually planning to make `${err}` a custom object with attributes containing the message and the actual exception as discussed above. This unfortunately turned out to be rather complicated. The main problem is that when there are continuable failures, or when a user keyword fails both in its body and in its teardown, there are multiple exceptions. In such cases `${err.exception}` needed to be a tuple or a custom object containing references to all occurred errors. This would be hard to implement properly and it would be very hard to use `${err.exception}` if you don't know what it actually contains.

Due to the problems explained above, we decided to make `${err}` just the error message. Robot does pretty good job in formatting the message nicely also if there are multiple failures so even non-technical users ought to be able to handle them. `${err}` being the exact same error message you see in the log file ought to make using this anyway pretty advanced functionality a bit easier in general. If more fine-grained control is needed, it can be done in Python level.

If you see big benefits in having access to the actual exception. please add a comment explaining concrete use cases. If there turns out to be lot of needs for it, we can still take another look at implementing it. You can also add :+1: and :-1: to this comment depending on do you like the current behavior of making `${err}` just the message or not.

**Comment by Tattoo** [comment]
Will there be changes to the output.xml schema because of this? Just trying to ponder the impact to RF Jenkins plugin

ping @asimell

**Comment by pekkaklarck** [comment]
Yes, there are going to be new nodes in output.xml. What exactly is still subject to change. Old nodes ought to not be changed.

**Comment by pekkaklarck** [comment]
The branch where this has been developed has now been merged to master. There were still few smallish things to be fixed and fine tuned and @yanne is currently working with them. After that there are these known open tasks:

- [ ] Decide should output.xml use `<tryblock>` or `<block type='try'>`. We currently use the former which means we have separate node types for `TRY`, `EXCEPT`, `ELSE` and `FINALLY`. The latter would mean we'd only get one new node type `<block>` with `type` attribute separating different block types.
- [ ] Update output.xml schema.
- [ ] Update schema compatibility tests to tests also TRY/EXCEPT.
- [ ] Test (and fix) using TRY/EXCEPT with templates. There's no clear use case for such functionality so we probably should explicitly prohibit it.

**Comment by pekkaklarck** [comment]
The TRY/EXCEPT functionality has been ready for some time, but there have still be some changes lately to the code, most importantly to various model objects representing TRY/EXCEPT structures. Now everything ought to be done for good, except for documentation. Once that is done and few other issues ready, we'll create alpha 1 to make it easier for everyone to test it. Hopefully the alpha is ready before the new year, but now we'll have a little Xmas break! :gift:

**Comment by pekkaklarck** [comment]
I've been writing documentation for this functionality and one thing I found somewhat annoying to explain is different matching strategies. I do believe it's useful that we support exact matching, glob patterns and regexps, but I'm not sure do we need to support `STARTS: Example` when its possible to use `GLOB: Example*`. I'm not entirely happy with the syntax that requires using a prefix like `GLOB: ValueError: *`. Could the pattern type be specified, for example, using a separate argument? It could be something like this:
```
TRY
    Something
EXCEPT    ValueError: *    type=glob
    Error Handling
END
```

This isn't any less verbose than what we currently have, but perhaps its a bit more readable. It would mean that if an `EXCEPT` has multiple messages, they'd all must use the same type. For example, we currently support
```
EXCEPT    GLOB: ValueError: *    GLOB: TypeError: *    REGEXP: Error \\d+
    Error Handling
```
and with a separate type argument that needed to be written like
```
EXCEPT    ValueError: *    TypeError: *    type=glob
    Error Handling
EXCEPT    Error \\d+    type=regexp
    Error Handling
```

I kind of like the separate  `type=` argument better than the type prefix, but don't see a big difference between them. The old `Run Keyword And  Except Error` keyword supporting prefixes could be a big enough reason to go with that approach. It would make it easier to replace usages of that keyword with the new native `TRY/EXCEPT` functionality.

Other alternatives I've had in mind:
- Some less verbose prefix like `*=`. A problem is that you cannot really know what `*= ValueError: *` means unless you read the documentation. A benefit of `GLOB:` is that it's pretty explicit.
- Make it a default to match the message first as an exact match and then as a glob match. Then we'd need a prefix only with regexps (`REGEXP: foo.*bar`) or when limiting matching only to exact match (`EQUALS: foo*bar`) or to glob patterns (`GLOB: foo*bar`).
- Use `/pattern/` with regexps. Globs could get their own syntax like `|pattern|` or we could make it a default to match the pattern first as an exact match and then as a glob as discussed above.
- Use something else than `EXCEPT` when with glob and regexps. Could be `EXCEPT GLOB`, `EXCEPT/GLOB` or just `GLOB`. I don't like having more reserved words in our syntax too much, and this isn't any less verbose than what we have currently.

The above discussion covered many related but different topics:
- Do we need functionality match only the start of the message or is it enough to have glob matches?
- Are type prefixes fine or should we try to come up with something better? 
- Should the default match strategy try both exact matching and glob pattern matching?

**Comment by pekkaklarck** [comment]
In my [previous comment](https://github.com/robotframework/robotframework/issues/3075#issuecomment-991658753) I discussed the `AS ${var}` variable and should it be just the error message or an object having attributes containing the message, the original error, traceback, etc. We decided to go with just the message for the following reasons:
1. It's easiest to understand for users.
2. There likely aren't many cases where the actual error or the traceback aren't needed. Generally it's better to have the logic in a library when such information is needed.
3. There are cases where multiple errors occur (continuable failures, failure both in a user keyword its teardown) and in such cases there couldn't really be `${err.exception}` attribute referring to the occurred error. We thought about wrapping multiple errors into some custom error object and using `${err.exceptions}` instead, but all these solutions would have just made both usage and implementation considerably harder.

**Comment by pekkaklarck** [comment]
As I wrote above, the `AS ${var}` variable nowadays contains just the error message. When using regular expressions and especially when the pattern contains capturing groups, it could be handy if the variable would be the match object. Consider this example:
```
TRY
    Something
EXCEPT    REGEXP: Error (\\d+) occurred    AS    ${match}
    Log    Error number was: ${match.group(1)}
END
```
Possibilities we could do:
- Always use the occurred error as the variable values. It's possible to use `Should Match Regexp` keyword separately to match regular expressions with capturing groups.
- Make the variable the match object with regular expressions and use error message with other matching strategies.
- Make the variable the match object with regular expressions if the pattern contains capturing groups, use error message in all other cases. Basically using capturing groups would mean opting in for getting mach objects.

**Comment by emanlove** [comment]
I like the readability of the type prefix over the separate argument. To me it puts the information about type right alongside of the expression for matching. I could see how it restricts the flexibility of the overall usage but for me that readability wins out.

On a different note a couple of items ..

First the combination of ignore_case argument with specifically the REGEX: matching, as shown here
```
TRY
    Something
EXCEPT    Error \\d+ occurred   type=regexp    ignore_case=True
    Error Handling
END
```
would seem problematic. I would think one doing regular expression would express the desire for ignore case within the regular expression it self. What if the regular expression specific said to include case information?

Second, could you use a syntax like
```
TRY
    Something
EXCEPT    ValueError: *    TypeError: *    type=glob
EXCEPT    Error \\d+    type=regexp
    Error Handling 1
EXCEPT    Alternative error
    Error handler 2
END
```
where you have consecutive exception lines (or ones with no body in between but just say comments) to allow for different exceptions to use the same keywords to handle various different errors?

**Comment by pekkaklarck** [comment]
It's true that regular expressions support case-insensitive matching. Unfortunately when you only can specify a pattern, not flags, you need to embed this information to the pattern like `(?i)Example`. It works but `ignore_case=True` is certainly easier to understand than `(?i)`. A bigger benefit would be that `ignore_case=True` would also work with other matching strategies.

I don't actually care too much about `GLOB:` vs. `type=glob´, but an option to make matching case-insensitive would be handy. `CASE-INSENSITIVE: Error` and especially `GLOB: CASE-INSENSITIVE: Error *` doesn't feel too good.

**Comment by pekkaklarck** [comment]
Decisions related to open questions listed above:
- We use prefixes for specifying the expected message pattern type but can consider other alternatives still before the release. Assuming prefixes would continue to work, we support for configuration like `regexp=True` could be added also in future versions without much backwards compatibility problems.
- We keep the `AS` variable as a string regardless the pattern type. Changing it to a regular match object with regexps if the pattern has groups can still be considered.
- We disallow using `RETURN`, `BREAK` and `CONTINUE` in `FINALLY`. `Pass Execution` will still work.

**Comment by pekkaklarck** [comment]
All open issues ought to now be covered. We can consider this issue done.

**Comment by pekkaklarck** [comment]
I reopen this due to a bug in parsing EXCEPT with AS with `data_only=False`. This bug was initially reported as #4208 by @d-biehl and there's PR #4209 by him to fix the problem as well.

**Comment by pekkaklarck** [comment]
The reported bug has now been fixed.

**Comment by pekkaklarck** [comment]
We decided to add a support to limit how many times WHILE loops are executed to avoid endless loops. This limit is specified using `limit=<value>` like this:

```
WHILE    ${condition}    limit=1 minute
    # ...
END
```

Now that we use `option=value` syntax with WHILE, it would be a good idea to use same syntax also with EXCEPT when controlling the pattern type. This is very late in RF 5.0 release cycle and we should already have the release out, but I believe it's better to change syntax still before the first release where TRY/EXCEPT is introduced than changing it later.

**Comment by pekkaklarck** [comment]
My thinking is that the pattern type would be specified using `pattern` option. If should follow the actual pattern or patterns, but be before the `AS` part. If there are multiple patterns, the `pattern` configuration would apply to them all. In practice it would work like this:
```
TRY
    Keyword
EXCEPT    Error: *    pattern=glob
   Handler 1
EXCEPT    (Another|Second) Error    Second [Pp]attern    pattern=regexp    AS    ${error}
    Handler 2
END
```

Things to decide:
- Should we use `pattern` or something else like `type`?
- Do we need to support matching only the start of the error message? We do that now, but just adding `*` at the end and using glob patterns is easy as well.
- What exact value to use with regular expressions? Possible we could support both `regex`, and `regexp`.
- What about case insensitivity? Should it be supported like `pattern=case-insensitive`  or should we add another option that could be used like `case_insensitive=True`?

**Comment by pekkaklarck** [comment]
Decisions:

**Comment by pekkaklarck** [comment]
New pattern type syntax was implemented in bb4b01ec560b641d8cfe48f141d989afa9a59154.

**Comment by pekkaklarck** [comment]
Documentation and few more smallish fixes done in above commit. Closing.
